agent/lib/kontena/workers/log_worker.rb

```diff
@@ -199,14 +205,16 @@ def on_container_event(topic, event)
       error exc
     end
 
-    # Close the log queue, stopping processing once empty. It cannot be restarted anymore.
-    def close
-      @queue.close
-    end
-
     def finalize
       stop_streaming if streaming?
-      close
+      pause_processing
+      if @queue.size > 0 && rpc_client
+        while @queue.size > 0 do
+          batch = @queue.shift(BATCH_SIZE * 10)
+          info "flushing #{batch.size} log items from queue"
```

Comment:
Good idea, but it doesn't work:

```
^CD, [2017-09-05T07:46:16.898600 #1] DEBUG -- Kontena::RpcClient: notify /containers/log_batch
D, [2017-09-05T07:46:16.950301 #1] DEBUG -- Kontena::RpcClient: notify /containers/log_batch
D, [2017-09-05T07:46:17.082422 #1] DEBUG -- Kontena::WebsocketClient: disconnected
I, [2017-09-05T07:46:17.084057 #1]  INFO -- Kontena::RpcClient: shutting down...
I, [2017-09-05T07:46:17.101454 #1]  INFO -- Kontena::Workers::ContainerHealthCheckWorker: stopped to check status from /ingress-lb.lb-1
I, [2017-09-05T07:46:17.106785 #1]  INFO -- Kontena::Workers::LogWorker: stop log streaming
I, [2017-09-05T07:46:17.386107 #1]  INFO -- Kontena::Workers::LogWorker: flushing 100 log items from queue
I, [2017-09-05T07:46:17.394775 #1]  INFO -- Kontena::Workers::LogWorker: flushing 100 log items from queue
I, [2017-09-05T07:46:17.394843 #1]  INFO -- Kontena::Workers::LogWorker: flushing 100 log items from queue
I, [2017-09-05T07:46:17.394873 #1]  INFO -- Kontena::Workers::LogWorker: flushing 100 log items from queue
I, [2017-09-05T07:46:17.394913 #1]  INFO -- Kontena::Workers::LogWorker: flushing 100 log items from queue
I, [2017-09-05T07:46:17.394951 #1]  INFO -- Kontena::Workers::LogWorker: flushing 100 log items from queue
I, [2017-09-05T07:46:17.395004 #1]  INFO -- Kontena::Workers::LogWorker: flushing 100 log items from queue
I, [2017-09-05T07:46:17.395030 #1]  INFO -- Kontena::Workers::LogWorker: flushing 100 log items from queue
I, [2017-09-05T07:46:17.395055 #1]  INFO -- Kontena::Workers::LogWorker: flushing 100 log items from queue
I, [2017-09-05T07:46:17.395079 #1]  INFO -- Kontena::Workers::LogWorker: flushing 100 log items from queue
I, [2017-09-05T07:46:17.395104 #1]  INFO -- Kontena::Workers::LogWorker: flushing 100 log items from queue
I, [2017-09-05T07:46:17.395127 #1]  INFO -- Kontena::Workers::LogWorker: flushing 100 log items from queue
I, [2017-09-05T07:46:17.395166 #1]  INFO -- Kontena::Workers::LogWorker: flushing 100 log items from queue
I, [2017-09-05T07:46:17.395189 #1]  INFO -- Kontena::Workers::LogWorker: flushing 100 log items from queue
I, [2017-09-05T07:46:17.395223 #1]  INFO -- Kontena::Workers::LogWorker: flushing 100 log items from queue
I, [2017-09-05T07:46:17.395247 #1]  INFO -- Kontena::Workers::LogWorker: flushing 90 log items from queue
```

I don't know what determines the shutdown order, but it looks l like it happens in exactly the wrong order, and these just get flushed out to `/dev/null` (dead actor mailboxes).

Reply:
I know it's not guaranteed to work but I thought that it's better to try flush those items from queue. Maybe WebsocketClient should keep connection open for a moment?

Reply:
AFAICT it's guaranteed to not work, and changing the actor shutdown order won't help either, in terms of trying to do any actor calls in the finalizers during the supervisor shutdown actor termination loop: https://github.com/celluloid/celluloid-supervision/blob/v0.20.5/lib/celluloid/supervision/container.rb#L137-L140

Once the `LogWorker` finalizer completes and the supervisor continues on to terminate the `RpcClient`/`WebsocketClient` actors, the `Celluloid::TerminationRequest is_a? Celluloid::SystemEvent`, so it jumps to the head of the mailbox, and any `async.notification` / `async.send_notification` calls that are still queued up in the terminating actor's mailbox will not get dispatched. I'm pretty sure that means that it's impossible to reliably flush out any queued up logs during the supervision shutdown.